pkg/provider/ngrok/client.go

```diff
@@ -0,0 +1,329 @@
+		Value:       ptr.To(string(value)),
+		Metadata:    ptr.To(string(metadataJSON)),
+		Description: ptr.To(psmd.Description),
+	})
+	return err
+}
+
+func (c *client) SecretExists(ctx context.Context, ref esv1.PushSecretRemoteRef) (bool, error) {
+	err := c.verifyVaultNameStillMatchesID(ctx)
+	if errors.Is(err, errVaultDoesNotExist) {
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
+
+	// Implementation for checking if a secret exists in ngrok
+	secret, err := c.getSecretByVaultIDAndName(ctx, c.getVaultID(), ref.GetRemoteKey())
+	if errors.Is(err, errVaultDoesNotExist) || errors.Is(err, errVaultSecretDoesNotExist) {
+		return false, nil
+	}
+
+	if err != nil {
+		return false, fmt.Errorf("error fetching secret: %w", err)
+	}
+
+	return (secret != nil), nil
+}
+
+// DeleteSecret deletes a secret from ngrok by its reference.
+func (c *client) DeleteSecret(ctx context.Context, ref esv1.PushSecretRemoteRef) error {
+	err := c.verifyVaultNameStillMatchesID(ctx)
+	if errors.Is(err, errVaultDoesNotExist) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	secret, err := c.getSecretByVaultIDAndName(ctx, c.getVaultID(), ref.GetRemoteKey())
+	if errors.Is(err, errVaultDoesNotExist) || errors.Is(err, errVaultSecretDoesNotExist) {
+		// If the secret or vault do not exist, we can consider it deleted.
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
+
+	if secret == nil {
+		return nil
+	}
+
+	return c.secretsClient.Delete(ctx, secret.ID)
+}
+
+func (c *client) Validate() (esv1.ValidationResult, error) {
+	// Validate the client can list secrets with a timeout. If we
+	// can list secrets, we assume the client is valid(API keys, URL, etc.)
+	iter := c.secretsClient.List(nil)
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	for iter.Next(ctx) {
+		return esv1.ValidationResultReady, nil
+	}
+
+	if iter.Err() != nil {
+		return esv1.ValidationResultError, fmt.Errorf("store is not allowed to list secrets: %w", iter.Err())
+	}
+
+	return esv1.ValidationResultReady, nil
+}
+
+func (c *client) GetSecret(ctx context.Context, _ esv1.ExternalSecretDataRemoteRef) ([]byte, error) {
+	// Implementation for getting a secret from ngrok
+	return nil, errWriteOnlyOperations
+}
+
+func (c *client) GetSecretMap(ctx context.Context, _ esv1.ExternalSecretDataRemoteRef) (map[string][]byte, error) {
+	// Implementation for getting a map of secrets from ngrok
+	return nil, errWriteOnlyOperations
+}
+
+func (c *client) GetAllSecrets(ctx context.Context, _ esv1.ExternalSecretFind) (map[string][]byte, error) {
+	// Implementation for getting all secrets from ngrok
+	return nil, errWriteOnlyOperations
+}
+
+func (c *client) Close(_ context.Context) error {
+	return nil
+}
+
+func (c *client) verifyVaultNameStillMatchesID(ctx context.Context) error {
+	if c.getVaultID() == "" {
+		return c.refreshVaultID(ctx)
+	}
+
+	vault, err := c.vaultClient.Get(ctx, c.getVaultID())
```

Comment:
you call c.getVaultID twice here which locks twice. instead store the vault id in a function variable and reuse it.